# 14.75 problem set 1

In [42]:
import pandas as pd
import numpy as np
import statsmodels.api as sm

## Question 1

Create data frame with 100 observations with

a) A variable `t` running from `1` to `100`

b) `alpha` which has value `3` for all 100 observations

c) Another variable `e_t`, a normal variable with mean 0 and standard deviation 1

d) `x_t` a random uniform variable over [0, 1]

In [19]:
SIZE = 100

In [29]:
df = pd.DataFrame({'t': np.arange(SIZE) + 1,
                   'alpha': 3,
                   'e_t': np.random.normal(size=SIZE),
                   'x_t': np.random.uniform(size=SIZE)})

e) An outcome variable `y_t` where `beta=2`, `y_t` = `alpha + beta * x_t + e_t`.

In [30]:
BETA = 2
df['y_t'] = df.alpha + BETA * df.x_t + df.e_t

f) Estimate `beta_hat`

  i. Test H0: `beta = 0`
  ii. Test H0: `beta = 1.2`

In [43]:
sm.OLS(endog=df.y_t,
       exog=sm.add_constant(df.x_t)).fit().summary()

/home/mghenis/miniconda3/lib/python3.7/site-packages/numpy/core/fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    y_t   R-squared:                       0.339
Model:                            OLS   Adj. R-squared:                  0.332
Method:                 Least Squares   F-statistic:                     50.15
Date:                Mon, 17 Feb 2020   Prob (F-statistic):           2.18e-10
Time:                        17:10:05   Log-Likelihood:                -148.45
No. Observations:                 100   AIC:                             300.9
Df Residuals:                      98   BIC:                             306.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.8360      0.209     13.584      0.000       2.422       3.250
x_t            2.4477      0.346      7.082      0.000       1.762       3.134
==============================================================================
Omnibus:                        0.246   Durbin-Watson:                   1.801
Prob(Omnibus):                  0.884   Jarque-Bera (JB):                0.063
Skew:                           0.056   Prob(JB):                        0.969
Kurtosis:                       3.052   Cond. No.                         4.13
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**95% CI includes neither 0 nor 1.2, so we can reject both nulls.**

g) Create `v_t` as a random normal variable with mean 0. Generate `q_t` as `q_t = x_t + 2*x_t^3 + v_t`.

What is the correlation between q and x?

Generate outcome variable `z_t` where `beta=2`, `gamma=3`, `z_t = alpha + beta * x_t + gamma * q_t + e_t`.

Estimate `beta_hat` from (misspecified) model:
`z_t = alpha + beta * x_t + u_t`

Test H0: `beta = 2`.

Discuss.

In [31]:
df['v_t'] = np.random.normal(size=SIZE)

In [32]:
df['q_t'] = df.x_t + 2 * np.power(df.x_t, 3) + df.v_t

Correlation:

In [41]:
df[['q_t', 'x_t']].corr().iloc[0, 1]

0.7530039456062745

In [33]:
GAMMA = 3

In [34]:
df['z_t'] = df.alpha + BETA * df.x_t + GAMMA * df.q_t + df.e_t

In [44]:
sm.OLS(endog=df.z_t,
       exog=sm.add_constant(df.x_t)).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    z_t   R-squared:                       0.625
Model:                            OLS   Adj. R-squared:                  0.621
Method:                 Least Squares   F-statistic:                     163.0
Date:                Mon, 17 Feb 2020   Prob (F-statistic):           1.44e-22
Time:                        17:14:41   Log-Likelihood:                -249.49
No. Observations:                 100   AIC:                             503.0
Df Residuals:                      98   BIC:                             508.2
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.7798      0.573      1.360      0.177      -0.358       1.918
x_t           12.1199      0.949     12.767      0.000      10.236      14.004
==============================================================================
Omnibus:                        2.401   Durbin-Watson:                   1.636
Prob(Omnibus):                  0.301   Jarque-Bera (JB):                2.416
Skew:                          -0.355   Prob(JB):                        0.299
Kurtosis:                       2.725   Cond. No.                         4.13
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Estimated `beta` = 12.1. 95% CI = [10.2, 14.004] so we can reject `beta=2`.
This is the result of omitted variable bias: `z_t` is being affected by `q_t` which is correlated with `x_t` in a nonlinear way that `x_t` can't soak up alone (if it were linear we'd expect `beta_hat = beta + gamma = 5`, rather than 12).**

In [45]:
sm.OLS(endog=df.z_t,
       exog=sm.add_constant(df[['x_t', 'q_t']])).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    z_t   R-squared:                       0.951
Model:                            OLS   Adj. R-squared:                  0.950
Method:                 Least Squares   F-statistic:                     933.7
Date:                Mon, 17 Feb 2020   Prob (F-statistic):           4.33e-64
Time:                        17:18:24   Log-Likelihood:                -148.05
No. Observations:                 100   AIC:                             302.1
Df Residuals:                      97   BIC:                             309.9
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9103      0.225     12.917      0.000       2.463       3.358
x_t            2.0979      0.526      3.989      0.000       1.054       3.142
q_t            3.1085      0.123     25.310      0.000       2.865       3.352
==============================================================================
Omnibus:                        0.746   Durbin-Watson:                   1.853
Prob(Omnibus):                  0.689   Jarque-Bera (JB):                0.391
Skew:                           0.131   Prob(JB):                        0.822
Kurtosis:                       3.160   Cond. No.                         9.77
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""